fms/modules/attention.py

```diff
@@ -58,15 +248,17 @@ def __init__(
         self.emb_v_per_head = emb_v
         self.p_dropout = p_dropout if p_dropout is not None else 0.0
         self.use_bias = use_bias
-        self.query = nn.Linear(
-            self.emb_dim, self.nheads * self.emb_kq_per_head, bias=use_bias
-        )
-        self.key = nn.Linear(
-            self.emb_dim, self.kvheads * self.emb_kq_per_head, bias=use_bias
-        )
-        self.value = nn.Linear(
-            self.emb_dim, self.kvheads * self.emb_v_per_head, bias=use_bias
+        self.fused = fused
+
+        self.in_proj: QKV = (FusedQKV if self.fused else UnfusedQKV)(
```

Comment:
I don't think we would switch back and forth within a single model. I wonder if it would be simpler to just add a FusesdMHA class distinct from MultiHeadAttention instead of adding this boolean switch.

Since the approach here adds three extra classes, just adding a FusedMHA might be less code and might be easier to read.

Reply:
adding a FusedMHA adds a lot of extra code for TP, which is why I told Josh to keep the changes to just this boolean instead, but at the same time, Fused/regular MHA are just different in this small part of the whole